### ISSUE TITLE
[BUG] DiskService.SetLength causes (random) exception

### ISSUE THREAD

**Issue by LennardF1989** [mbdavid/LiteDB#1623]
We're currently using LiteDb to store about 350k small objects in a Collection. This is inside a .NET Core 3.1 ASP.NET Application.

At random times, but pretty consistently in every run, the following Exception is thrown. Even when the dataset is as small as 5000, to stay under the BulkInsert default Bulk-size.

```
IOException: The requested operation cannot be performed on a file with a user-mapped section open. : 'Database-log.db'
System.IO.FileStream.SetLengthCore(long value)
System.IO.FileStream.SetLengthInternal(long value)
System.IO.FileStream.SetLength(long value)
LiteDB.Engine.DiskService.SetLength(long length, FileOrigin origin)
LiteDB.Engine.WalIndexService.Clear()
```
Whenever we switch to Shared, the issue obviously goes away because a log-file is no longer used, but the issues reported by others in the last few hours start occurring.

Upon investigating, I tried a bunch of different things, but eventually decided to just "timeout" SetLength and try again using the following code:
```
            int attempt = 1;

            tryAgain:

            try
            {
                stream.SetLength(length);
            }
            catch
            {
                Debug.WriteLine($"SetLength #{attempt} failed, retrying in 100ms...");

                Task.Delay(100).Wait();

                attempt++;

                goto tryAgain;
            }
```
I've run the 350k BulkImport with this piece of code in the `DiskService.SetLength` multiple times, and never seen it reach a second attempt. This leads me to conclude two things:
- A race condition is happening inside the code of LiteDb, a Reader stream is being used at a time where the DiskService wants to write.
- An external factor like a virusscanner is at play here.

For the first, I cannot say for sure, but I'm not seeing weird things in the code, it all seems properly thread-safe.

For the second, in an attempt to figure out the problem, I made a test-project which does exactly the same things in the same way. However, even after multiple runs, I cannot reproduce the issue at all, which is frustrating. If a virusscanner was too noisy, my test-bed would break. The test-bed I attached is an exact replica of the current code-base where we have this problem, so it's weird it happens in the "real deal", but not the test-version...

I hope you find this useful and can figure out a potential reason why this happens. Attached is my test-bed for this particular issue containing a Console-app and a Web-app mirroring the exact situation where have this issue on. I obviously cannot supply the realData.json, but I can confirm the data itself does not matter for the actual Exception to happen.

[LiteDBTester.zip](https://github.com/mbdavid/LiteDB/files/4453008/LiteDBTester.zip)

PS. It looks like that using Insert instead of BulkInsert can trigger the Exception faster.

**Comment by lbnascimento** [comment]
@LennardF1989 I ran your test project and I couldn't reproduce the issue either. There's probably something different in how the real code is using LiteDB - maybe its datafile is in a place that "activates" your anti-virus, while the test project is not.

If there was any way I could run (or at least take a look at) the real code, that could be helpful.

**Comment by LennardF1989** [comment]
@lbnascimento I figured it out. Knowing what I know now, it is so obvious...

I noticed that from one commit to another things kept working as it should. So I started going back in time until it broke again. Turns out... sigh... it was my TortoiseGit trying to generate a status icon in Windows Explorer for the file because it got add to the repository.

As soon as I told git to delete the file and stop tracking it, LiteDB worked perfectly, even with 350k Insert-calls in a row. As soon as I added the file back to enable tracking, the issue returned.

In the end, it was not a virusscanner, but my trusty 'ol TortoiseGit, whom I never suspected at all.

I'll close this issue, but let it be a warning for all: Something else can be affecting your files as well! In my case, it was my least suspected and useful tool that stabbed me in the back :(
